### ISSUE TITLE
loss calculation for multi-dim samples in distributed strategy

### ISSUE THREAD

**Issue by llodds** [tensorflow/tensorflow#48743]
Hi, I was reading the distributed strategy and found one issue in the official doc
https://www.tensorflow.org/tutorials/distribute/custom_training

> If labels is multi-dimensional, then average the per_example_loss across the number of elements in each sample. For example, if the shape of predictions is (batch_size, H, W, n_classes) and labels is (batch_size, H, W), you will need to update per_example_loss like: per_example_loss /= tf.cast(tf.reduce_prod(tf.shape(labels)[1:]), tf.float32)

Since the last dimension is automatically reduced in "NONE" mode (https://github.com/tensorflow/tensorflow/issues/27190), then the manual reduce formula for the distributed.strategy should be
per_example_loss /= tf.cast(tf.reduce_prod(tf.shape(labels)[1:**-1**]), tf.float32) # notice the addition of -1 in shape() to remove the accumulation for the last axis

I don't know which place I should put this post to, but please update the relevant people who is in charge of the doc. Thanks.

**Comment by MarkDaoust** [comment]
Hi, 

The source doc for this page is here:

https://github.com/tensorflow/docs/blob/master/site/en/tutorials/distribute/custom_training.ipynb

I need a few minutes to think about that loss reduction issue.

**Comment by llodds** [comment]
I did a test to verify this.

**Comment by MarkDaoust** [comment]
[Note: if you post the notebook to GitHub without zipping it then people can open it directly in google Colab without leaving the browser by replacing `github.com` with `https://colab.sandbox.google.com/github/`]

I see your point from #27190 that this is how it works.  I just really don't like it 😅.

```
(batch_size, H, W, n_classes) and labels is (batch_size, H, W), you will need to update per_example_loss like: per_example_loss /= tf.cast(tf.reduce_prod(tf.shape(labels)[1:]), tf.float32)
```

Anyway, the problem with just changing this line is that while the loss classes are [usually] a good way to implement loss functions, any callable can used.  So we have to say something generic, and not assume that anyone doing this is using one of the builtin losses with `Reduction.NONE`. Also, for this specific example `(batch_size, H, W, n_classes)` with `tf.keras.losses.SparseCategoricalCrossentropy` (the code just below) it's the `n_classes` dimension that gets reduced making the current code correct in that case.

So I think if we want to fix this, the right approach is to add:

```
Warning: `losses` typically take the average across the the last axis of the input  (even if you pass `reduction=Reduction.NONE` when creating a loss object). For `losses.CategoricalCrossEntropy` this reduces the `n_classes` dimension which is typically what you want. For losses that are calculated pointwise, like mean squared error or binary crossentropy `[batch, W, H]` will be reduced to `[batch, W]`, unless  you include a dummy axis like `[batch, W, H, 1]`.

**Comment by llodds** [comment]
I see your original intention now, to cover something more generic. But please add this **caution** to warn people like me who is a TF newbie and did not realize loss functions already reduce last axis even when Reduction is NONE (the naming "NONE" is really confusing...) Thanks!

I will make sure next time when I submit a ticket I will submit a notebook rather than zip it...
